bake/types.go

```diff
@@ -0,0 +1,51 @@
+package bake
+
+import (
+	"strings"
+
+	"github.com/docker/buildx/util/buildflags"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+)
+
+type ExportEntry struct {
+	Type        string            `json:"type"`
+	Attrs       map[string]string `json:"attrs,omitempty"`
+	Destination string            `json:"dest,omitempty"`
+}
+
+func (e *ExportEntry) FromCtyValue(in cty.Value) error {
+	// Attempt to read as a cty.Map(cty.String) first as that's our primary form.
+	if conv, err := convert.Convert(in, cty.Map(cty.String)); err == nil {
+		// Build the csv string for parsing. We do this mostly because
+		// ParseExports expects that format and it's easier to do that then reimplement it
+		// in multiple locations.
+		var sb strings.Builder
+		for name, val := range conv.AsValueMap() {
```

Comment:
We shouldn't encode objects to `csv` string. That is backwards as the (typed) object is what we need.

If you want to reduce duplication then there can be generic `csv -> map[k]v` convert and then map to struct, although not sure if worth it.

Reply:
Updated the prototype so the `map[k]v` version is the canonical form.

To reduce duplication, I've created a separate `ExportEntry` type with the same attributes. This type will implement `encoding.TextUnmarshaler` and `json.Unmarshaler` to control the marshaling and unmarshaling process in a consistent way. The type also has a `ToPB()` method to convert the type to the controller API version.

The csv parsing has been extracted into `UnmarshalText` and the original `ParseExports` method has been changed to be a loop that does this:

```
var out ExportEntry
if err := out.UnmarshalText([]byte(s)); err != nil {
    return nil, err
}
outs = append(outs, out.ToPB())
```

The a `FromCtyValue` method has been created so we can still detect it. But, we now only need one implementation that uses `UnmarshalText` and JSON unmarshaling and we should be able to reuse this implementation for each of these types.